effects/push-cache/default.nix

```diff
@@ -0,0 +1,160 @@
+              example = "[ pkgs.hello ]";
+            };
+            secretName = lib.mkOption {
+              type = lib.types.str;
+              description = ''
+                Name of the HerculesCI secret. See [HerculesCI docs](https://docs.hercules-ci.com/hercules-ci-agent/secrets-json).
+                The secrets "data" field should contain given data:
+
+                ```
+                  "data": {
+                    "name": "my-cache-name",
+                    "token": "ey536428341723812",
+                    "endpoint": "https://my-cache-name.com"
+                  }
+                ```
+
+                The "endpoint" field is needed for Attic cache. With Cachix cache the "endpoint" field is not read and can be absent.
+              '';
+            };
+            branches = lib.mkOption {
+              type = with lib.types; nullOr (listOf str);
+              default = null;
+              description = ''
+                Branches on which we'd like to execute the effect. Set to `null` to execute on all branches.
+              '';
+            };
+          };
+        }));
+      };
+    };
+  };
+
+  config = let
+    # file with all the package paths written line by line
+    # nixpkgs -> [derivation] -> derivation
+    packagesFile = pkgs: packages:
+      pkgs.writeText "pushed-paths" (lib.strings.concatStringsSep "\n"
+        (builtins.map builtins.toString packages));
+
+    mkAtticPushEffect = { cacheOptions, branch, }:
+      withSystem "x86_64-linux" ({ hci-effects, pkgs, ... }:
+        let
+          pushEffect = hci-effects.mkEffect {
+            inputs = [ config.push-cache-effect.attic-client-pkg ];
+            secretsMap = { token-file = "${cacheOptions.secretName}"; };
+            userSetupScript = ''
+              attic login \
+                server-name \
+                $(readSecretString token-file .endpoint) \
+                $(readSecretString token-file .token)
+            '';
+            effectScript = ''
+              cat ${
+                packagesFile pkgs cacheOptions.packages
+              } | xargs -s 4096 attic push server-name:$(readSecretString token-file .name)
+            '';
+          };
+        in hci-effects.runIf ((cacheOptions.branches == null)
+          || (builtins.elem branch cacheOptions.branches)) pushEffect);
+
+    mkCachixPushEffect = { cacheOptions, branch, }:
+      withSystem "x86_64-linux" ({ hci-effects, pkgs, ... }:
+        let
+          pushEffect = hci-effects.mkEffect {
+            inputs = [ config.push-cache-effect.cachix-pkg ];
+            secretsMap = { token-file = "${cacheOptions.secretName}"; };
+            userSetupScript = ''
+              cachix authtoken $(readSecretString token-file .token)
+            '';
+            effectScript = ''
+              cat ${
+                packagesFile pkgs cacheOptions.packages
+              } | cachix push $(readSecretString token-file .name)
```

Comment:
Cache name doesn't need to be a secret. The whole "server-name:cache-name" argument should be taken as one string from the `name` option.